servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -625,6 +682,42 @@ private void handleResolveDone0(final Future<DnsAnswer<T>> addressFuture) {
                 }
             }
 
+            private void reportResolutionFailed(@Nullable final DnsResolutionObserver resolutionObserver,
+                                                final Throwable cause) {
+                if (resolutionObserver == null) {
+                    return;
+                }
+                try {
+                    resolutionObserver.resolutionFailed(cause);
+                } catch (Throwable unexpected) {
+                    LOGGER.warn("Unexpected exception from {} while reporting DNS resolution failure",
+                            resolutionObserver, unexpected);
+                }
+            }
+
+            private void reportResolutionResult(@Nullable final DnsResolutionObserver resolutionObserver,
+                                                final DnsAnswer<T> dnsAnswer,
+                                                @Nullable final List<ServiceDiscovererEvent<T>> events) {
+                if (resolutionObserver == null) {
+                    return;
+                }
+                final ResolutionResult result;
+                final int ttl = (int) NANOSECONDS.toSeconds(dnsAnswer.ttlNanos());
+                if (events == null) {
+                    result = new DefaultResolutionResult(dnsAnswer.answer().size(), ttl, 0, 0);
+                } else {
+                    final int becameActive = (int) events.stream().filter(ServiceDiscovererEvent::isAvailable).count();
```

Comment:
Instead of reiterating the events list can we have `calculateDifference()` method accept a `BiFunction` to report the numbers?